<img src="https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Módulo** | Análise de Dados: Fundamentos de Estatística
Caderno de **Exercícios**<br> 
Professor [André Perez](https://www.linkedin.com/in/andremarcosperez/)

---

# **Tópicos**

<ol type="1">
  <li>Média e Variância;</li>
  <li>Ordem e Posição;</li>
  <li>Correlação.</li>
</ol>

---

# **Exercícios**

## 1\. Tráfego de São Paulo

Neste exercício, vamos continuar a analisar os dados de mobilidade urbana da cidade de São Paulo. A base de dados está neste [link](https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv) e é uma copia do dado original, presente neste [link](https://archive.ics.uci.edu/ml/datasets/Behavior+of+the+urban+traffic+of+the+city+of+Sao+Paulo+in+Brazil). A base de dados contem a quantidade de acidentes ocorridos na cidade entre 14/12/09 e 18/12/09, das 07:00h ás 20:00h, agregados em intervalos de 30 minutos.

In [1]:
!wget -q "https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv" -O traffic.csv

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('traffic.csv', sep=';')

In [4]:
df.head()

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
0     1           0             0               0                0   
1     2           0             0               0                0   
2     3           0             0               0                0   
3     4           0             0               0                0   
4     5           0             0               0                0   

   running_over  fire_vehicles  occurrence_freight  \
0             0              0                   0   
1             0              0                   0   
2             0              0                   0   
3             0              0                   0   
4             0              0                   0   

   incident__dangerous_freight  lack_electricity  fire  flooding  \
0                            0                 0     0         0   
1                            0                 0     0         0   
2                            0                 0     0         0   
3                            0                 0     0         0   
4                            0                 0     0         0   

   manifestations  defect_trolleybuses  tree_road  semaphore_off  \
0               0                    0          0              0   
1               0                    0          0              0   
2               0                    0          0              0   
3               0                    0          0              0   
4               0                    0          0              0   

   intermittent_emaphore slowness_traffic_%  
0                      0                4,1  
1                      0                6,6  
2                      0                8,7  
3                      0                9,2  
4                      0               11,1

- **Exemplo**: (para relembrar)

O código abaixo extrai a 13ª linha do arquivo que representa a 14ª meia hora do dia 14/12/09, contadas a partir das 07:00h, ou seja, todos os incidentes ocorridos na cidade no dia em questão entre as 13:30h e 14:00h. Vemos, por exemplo, que dois ônibus e dois caminhões ficaram paradados nas ruas da cidade, atrapalhando o tráfego.

In [5]:
df.iloc[[13]]

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
13    14           2             2               0                1   

    running_over  fire_vehicles  occurrence_freight  \
13             0              0                   0   

    incident__dangerous_freight  lack_electricity  fire  flooding  \
13                            0                 0     0         0   

    manifestations  defect_trolleybuses  tree_road  semaphore_off  \
13               0                    0          0              0   

    intermittent_emaphore slowness_traffic_%  
13                      0                4,9

In [18]:
df.shape

(135, 18)

### **1.1. Agregação** 

Neste primeira atividade, você deve gerar um array NumPy por dia. Para cada array você deve somar todos os incidentes que aconteceram naquela meia hora. Sendo assim, cada array deve ter 27 posições, cada qual com a soma dos incidentes daquela meira hora.

**Dica**: Você deve remover a primeira e a última coluna.

**Dica**: Os arrays devem ter os seguintes valores:

```python
14: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 5, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 2, 3]
15: [3, 0, 0, 0, 2, 4, 2, 1, 3, 0, 2, 0, 2, 3, 2, 2, 4, 7, 6, 3, 0, 1, 1, 5, 4, 6, 0]
16: [1, 1, 2, 4, 4, 5, 3, 2, 4, 2, 1, 1, 3, 6, 2, 0, 2, 8, 4, 6, 4, 1, 2, 7, 15, 4, 0]
17: [4, 1, 1, 0, 0, 0, 2, 3, 2, 7, 2, 1, 7, 3, 6, 4, 3, 2, 2, 3, 2, 5, 2, 0, 1, 4, 15]
18: [10, 1, 1, 5, 5, 7, 3, 4, 3, 1, 2, 3, 2, 2, 6, 2, 2, 4, 2, 3, 0, 3, 3, 2, 3, 6, 0]
```

In [8]:
import pandas as pd
import numpy as np

#Primeiro carregar o arquivo CSV
df = pd.read_csv('traffic.csv', sep=';')

#Depois retirar as colunas desnecessárias
df = df.drop(columns=['hour', 'slowness_traffic_%'])

#Estabelecer quais são as matrizes e somá-las
dia_14 = np.array(df.iloc[0:27].T.values.reshape(16, 27))
dia_14_sum = np.sum(dia_14, axis=0)

dia_15 = np.array(df.iloc[27:54].T.values.reshape(16, 27))
dia_15_sum = np.sum(dia_15, axis=0)

dia_16 = np.array(df.iloc[54:81].T.values.reshape(16, 27))
dia_16_sum = np.sum(dia_16, axis=0)

dia_17 = np.array(df.iloc[81:108].T.values.reshape(16, 27))
dia_17_sum = np.sum(dia_17, axis=0)

dia_18 = np.array(df.iloc[108:135].T.values.reshape(16, 27))
dia_18_sum = np.sum(dia_18, axis=0)

#Imprimir os dados
print(f'Dia 14:\n{dia_14_sum}')
print(f'Dia 15:\n{dia_15_sum}')
print(f'Dia 16:\n{dia_16_sum}')
print(f'Dia 17:\n{dia_17_sum}')
print(f'Dia 18:\n{dia_18_sum}')

#Não faço ideia do motivo pelo qual os três últimos vieram separados, eu literalmente copiei e colei e só mudei os números depois de fazer o primeiro.

Dia 14:
[0 0 0 0 0 0 0 0 0 0 0 1 0 5 2 0 0 0 0 0 1 0 0 0 2 2 3]
Dia 15:
[3 0 0 0 2 4 2 1 3 0 2 0 2 3 2 2 4 7 6 3 0 1 1 5 4 6 0]
Dia 16:
[ 1  1  2  4  4  5  3  2  4  2  1  1  3  6  2  0  2  8  4  6  4  1  2  7
 15  4  0]
Dia 17:
[ 4  1  1  0  0  0  2  3  2  7  2  1  7  3  6  4  3  2  2  3  2  5  2  0
  1  4 15]
Dia 18:
[10  1  1  5  5  7  3  4  3  1  2  3  2  2  6  2  2  4  2  3  0  3  3  2
  3  6  0]


### **1.2. Métricas** 

Para cada array você deve calcular as seguintes métricas: 

 - **média**;
 - **desvio padrão**.

In [23]:
#Aqui minha tentativa de organizar tudo em apenas uma linha

print(f'Média do dia 14:', np.mean(dia_14_sum),'e desvio padrão:', np.std(dia_14_sum))
print(f'Média do dia 15:', np.mean(dia_15_sum),'e desvio padrão:', np.std(dia_15_sum))
print(f'Média do dia 16:', np.mean(dia_16_sum),'e desvio padrão:', np.std(dia_16_sum))
print(f'Média do dia 17:', np.mean(dia_17_sum),'e desvio padrão:', np.std(dia_17_sum))
print(f'Média do dia 18:', np.mean(dia_18_sum),'e desvio padrão:', np.std(dia_18_sum))

Média do dia 14: 0.5925925925925926 e desvio padrão: 1.1944085553034889
Média do dia 15: 2.3333333333333335 e desvio padrão: 2.0184335693983275
Média do dia 16: 3.4814814814814814 e desvio padrão: 3.047407767514698
Média do dia 17: 3.037037037037037 e desvio padrão: 3.0487578707046925
Média do dia 18: 3.1481481481481484 e desvio padrão: 2.1892617342417258


### **1.3. Interpretação** 

Baseado nos resultados da questão 1.2, responda:

> Qual dia apresenta a maior média de acidentes por meia hora?

**Resposta:** O dia 16 apresenta a maior média de acidentes, com 3.48 acidentes por período de meia hora. Embora o dia 15 possua um número similar de acidentes, sua média é menor por conta da variação também menor. O dia 16 possui valores que variam entre 0 e 15 acidentes em cada período, e esses 'outliers' contribuem para o aumento da média.

> Qual dia apresenta a menor variação de acidentes por meia hora?

**Resposta:** O dia 14 apresenta a menor variação de acidentes por meia hora, sendo valores baixos durante todo o dia(07:00 - 20:00). O número de acidentes se manteve baixo durante todo o período, e os valores discrepantes ainda estão próximos da média.

---